File: packages/core/src/utils/author-error-catalog.ts
/**
 * R7 结构化错误目录（368 号契约层，二轮 P2；355 号 §4 R7）。
 *
 * 唯一数据文件 = `data/author-errors.json`（双端共享：TS import JSON；
 * Rust `include_str!` 同一文件）——author-report、通知中心、Doctor 三面
 * 共用同一错误→分类→下一步的映射。
 *
 * 条目 shape：`{code, severity(auto-handled|needs-review|must-handle),
 * patterns[], messageZh/En, nextActionZh/En}`；patterns 为小写关键词
 * （对 `${event} ${message}` 小写串做 includes 匹配）；末条 unknown-error
 * 无 patterns 作为兜底（must-handle）。
 *
 * 双端：`engine-rs/src/utils/author_error_catalog.rs` 1:1 镜像；行为
 * golden = `src/__tests__/golden/author-error-catalog-vectors.json`
 * （TS 断言 `golden-author-error-catalog.test.ts`；Rust 差分
 * `tests/golden_author_error_catalog_diff.rs`）。
 */

import catalogData from "../data/author-errors.json";

export type AuthorErrorSeverity = "auto-handled" | "needs-review" | "must-handle";

export interface AuthorErrorEntry {
  readonly code: string;
  readonly severity: AuthorErrorSeverity;
  readonly patterns: ReadonlyArray<string>;
  readonly messageZh: string;
  readonly messageEn: string;
  readonly nextActionZh: string;
  readonly nextActionEn: string;
}

export interface AuthorErrorCatalog {
  readonly version: number;
  readonly errors: ReadonlyArray<AuthorErrorEntry>;
}

const CATALOG: AuthorErrorCatalog = catalogData as AuthorErrorCatalog;

export function loadAuthorErrorCatalog(): AuthorErrorCatalog {
  return CATALOG;
}

export interface ResolvedAuthorError {
  /** 命中条目 code；兜底条目为 "unknown-error"。 */
  readonly code: string;
  readonly severity: AuthorErrorSeverity;
  readonly message: string;
  readonly nextAction: string;
}

/** 目录查找：patterns 顺序匹配（haystack 小写 includes）；无命中走兜底条目。 */
export function resolveAuthorError(
  event: string,
  message: string,
  language: "zh" | "en" = "zh",
): ResolvedAuthorError {
  const isEn = language === "en";
  const haystack = `${event} ${message}`.toLowerCase();
  const errors = CATALOG.errors;
  let matched: AuthorErrorEntry | undefined;
  for (const entry of errors) {
    if (entry.patterns.some((pattern) => haystack.includes(pattern.toLowerCase()))) {
      matched = entry;
      break;
    }
  }
  if (!matched) {
    matched = errors.find((entry) => entry.patterns.length === 0);
  }
  const fallback = matched ?? errors[errors.length - 1]!;
  return {
    code: fallback.code,
    severity: fallback.severity,
    message: isEn ? fallback.messageEn : fallback.messageZh,
    nextAction: isEn ? fallback.nextActionEn : fallback.nextActionZh,
  };
}

/** 按 severity 取下一步建议（未知 severity 走兜底条目）。 */
export function nextActionFor(severity: AuthorErrorSeverity, language: "zh" | "en" = "zh"): string {
  const entry = CATALOG.errors.find((item) => item.severity === severity);
  const isEn = language === "en";
  if (entry) return isEn ? entry.nextActionEn : entry.nextActionZh;
  const fallback = CATALOG.errors[CATALOG.errors.length - 1]!;
  return isEn ? fallback.nextActionEn : fallback.nextActionZh;
}

File: engine-rs/src/utils/author_error_catalog.rs
//! R7 结构化错误目录（368 号契约层，二轮 P2）。
//!
//! TS 真源：`packages/core/src/utils/author-error-catalog.ts`；数据文件
//! `packages/core/src/data/author-errors.json`（`include_str!` 同一文件，
//! 双端共享同一错误→分类→下一步映射，author-report/通知中心/Doctor 三面
//! 共用）。行为 golden = `golden/author-error-catalog-vectors.json`。
//!
//! 移植纪律：patterns 匹配大小写不敏感（haystack 与 pattern 均小写
//! includes）；目录顺序即优先级；无命中走末条兜底（patterns 为空）。

use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuthorErrorEntry {
    pub code: String,
    pub severity: String,
    #[serde(default)]
    pub patterns: Vec<String>,
    pub message_zh: String,
    pub message_en: String,
    pub next_action_zh: String,
    pub next_action_en: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuthorErrorCatalog {
    pub version: i64,
    pub errors: Vec<AuthorErrorEntry>,
}

pub const AUTHOR_ERRORS_JSON: &str =
    include_str!("../../../packages/core/src/data/author-errors.json");

pub fn load_author_error_catalog() -> AuthorErrorCatalog {
    serde_json::from_str(AUTHOR_ERRORS_JSON).expect("author-errors.json should be valid")
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedAuthorError {
    pub code: String,
    pub severity: String,
    pub message: String,
    pub next_action: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CatalogLanguage {
    Zh,
    En,
}

/// 目录查找：patterns 顺序匹配（haystack 小写 includes）；无命中走兜底条目。
pub fn resolve_author_error(
    event: &str,
    message: &str,
    language: CatalogLanguage,
) -> ResolvedAuthorError {
    let catalog = load_author_error_catalog();
    let is_en = language == CatalogLanguage::En;
    let haystack = format!("{} {}", event, message).to_lowercase();
    let matched = catalog
        .errors
        .iter()
        .find(|entry| {
            entry
                .patterns
                .iter()
                .any(|pattern| haystack.contains(&pattern.to_lowercase()))
        })
        .or_else(|| catalog.errors.iter().find(|entry| entry.patterns.is_empty()))
        .or_else(|| catalog.errors.last());
    let entry = match matched {
        Some(entry) => entry,
        None => {
            return ResolvedAuthorError {
                code: "unknown-error".to_string(),
                severity: "must-handle".to_string(),
                message: if is_en { "The run failed and needs manual attention.".to_string() } else { "任务失败，需人工介入。".to_string() },
                next_action: if is_en { "Check Doctor diagnostics and logs, fix the cause, then re-run.".to_string() } else { "查看 Doctor 诊断与日志定位原因，修复后重跑。".to_string() },
            };
        }
    };
    ResolvedAuthorError {
        code: entry.code.clone(),
        severity: entry.severity.clone(),
        message: if is_en { entry.message_en.clone() } else { entry.message_zh.clone() },
        next_action: if is_en { entry.next_action_en.clone() } else { entry.next_action_zh.clone() },
    }
}

/// 按 severity 取下一步建议（未知 severity 走兜底条目）。
pub fn next_action_for(severity: &str, language: CatalogLanguage) -> String {
    let catalog = load_author_error_catalog();
    let is_en = language == CatalogLanguage::En;
    match catalog.errors.iter().find(|entry| entry.severity == severity) {
        Some(entry) => {
            if is_en { entry.next_action_en.clone() } else { entry.next_action_zh.clone() }
        }
        None => {
            let fallback = catalog.errors.last().expect("catalog has fallback entry");
            if is_en { fallback.next_action_en.clone() } else { fallback.next_action_zh.clone() }
        }
    }
}
